src/main/java/org/codeforamerica/messaging/services/MessageService.java

```diff
@@ -17,29 +15,39 @@ public class MessageService {
     private final SmsService smsService;
     private final EmailService emailService;
     private final MessageRepository messageRepository;
+    private final TemplateRepository templateRepository;
 
 
-    public MessageService(SmsService smsService, EmailService emailService, MessageRepository messageRepository) {
+    public MessageService(SmsService smsService,
+            EmailService emailService,
+            MessageRepository messageRepository,
+            TemplateRepository templateRepository) {
         this.smsService = smsService;
         this.emailService = emailService;
         this.messageRepository = messageRepository;
+        this.templateRepository = templateRepository;
     }
 
     public Message sendMessage(MessageRequest messageRequest) {
+        Template template = templateRepository.findFirstByName(messageRequest.getTemplateName());
+        if (template == null) {
+            throw new RuntimeException("Template not found with the name provided");
+        }
+
         Message message = Message.builder()
-                .subject(messageRequest.getSubject())
-                .body(messageRequest.getBody())
                 .toPhone(messageRequest.getToPhone())
                 .toEmail(messageRequest.getToEmail())
+                .template(template)
                 .build();
-        SmsMessage sentSmsMessage;
-        EmailMessage sentEmailMessage;
+
+        String subject = template.buildSubjectFromTemplate(messageRequest.getTemplateParams());
```

Comment:
I feel that the following would be clearer:
```
String subject = template.build(messageRequest.getSubject(), messageRequest.getTemplateParams());
String body = template.build(messageRequest.getBody(), messageRequest.getTemplateParams());
```

Reply:
The way I have it, `messageRequest` only has the name of the template, not the subject and body. And it would seem odd to me to call `template.getSubject()` only to pass it into `template.build()`. How about a change like this?
```
String subject = template.build(Template::getSubject, messageRequest.getTemplateParams());
String body = template.build(Template::getBody, messageRequest.getTemplateParams());
```

Or the lambda syntax instead of method reference?
```
String subject = template.build(template -> template.getSubject(), messageRequest.getTemplateParams());
String body = template.build(template -> template.getBody(), messageRequest.getTemplateParams());
```

Reply:
Silly me, still extracting subject and body from the `MessageRequest`. I like the method reference syntax better as it is concise.